### PR TITLE
Refactor how AOTAutograd backends are defined

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89773
* #89672
* #89671
* #89670
* #89669
* __->__ #89736

There was a lot of strangeness in how AOTAutograd backends were previously defined. This refactor replaces the strangeness with something simple and straightforward. The improvements:

- There is no longer a footgun aot_autograd "backend" which doesn't actually work. No more mistyping `torch._dynamo.optimize("aot_autograd")` when you meant "aot_eager"
- Deleted aot_print because it's annoying and anyway there's no uses of it
- Instead of having BOTH the backend Subgraph and AotAutogradStrategy, there is now only an aot_autograd function which takes the kwargs to configure AOTAutograd, and then gives you a compiler function that does AOTAutograd given those kwargs. Easy.
- The primary downside is that we are now eagerly populating all of the kwargs, and that can get us into import cycle shenanigans. Some cycles I resolved directly (e.g., we now no longer manually disable the forward function before passing it to aot_autograd; aot_autograd it does it for us), but for getting inductor decompositions I had to make it take a lambda so I could lazily populate the decomps later.

New code is 130 lines shorter!

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -1315,7 +1314,6 @@ class CompiledFunction(torch.autograd.Function):
         fw_metadata = _fw_metadata
 
         @staticmethod
-        @disable_torchdynamo
```
Comment (anjali411) [inline functorch/_src/aot_autograd.py:1318]: why were we disabling dynamo earlier and not anymore?

File: torch/_dynamo/optimizations/training.py [hunk 2]
```diff
@@ -22,6 +31,62 @@
...
+        if callable(kwargs.get("decompositions")):
+            kwargs["decompositions"] = kwargs["decompositions"]()
+
+        # TODO: stop monkeypatching here (without even cleaning up, UGH!)
+        functorch.compile.config.use_functionalize = True
+        functorch.compile.config.use_fake_tensor = True
+
+        force_compile_tiny_graphs = kwargs.pop("force_compile_tiny_graphs", False)
+
+        if count_calls(gm.graph) < 2 and not force_compile_tiny_graphs:
```
Comment (anjali411) [inline torch/_dynamo/optimizations/training.py:48]: as per the comment below for decomps, should we always force compilation?

**Review by anjali411 — APPROVED** [review]

File: functorch/_src/aot_autograd.py (hunk 1 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1318]: I moved the disable. Previously, we manually disabled dynamo on the forwards function we return. Now, dynamo is responsible for disabling itself on the returned compiled function. See https://github.com/pytorch/pytorch/pull/89736/files#diff-0b094ea719a9acd16c316b7ec6975f6c9825398952e9b3a67f180d20ac592d47R82

I wasn't really planning to do this change but import cycles were annoying

File: torch/_dynamo/optimizations/training.py (hunk 2 above)
Comment (anjali411) [reply torch/_dynamo/optimizations/training.py:48]: hmm yeah. I like the idea to force recompilation in debug mode anyway.

Comment (ezyang) [reply torch/_dynamo/optimizations/training.py:48]: @Chillee you cool with this idea?

File: torch/_dynamo/optimizations/training.py [hunk 3]
```diff
@@ -86,107 +151,28 @@ def raise_or_warn(reason):
...
-    def candidate(self):
-        from functorch.compile import ts_compile
-
-        return BACKENDS["aot_autograd"](
-            self.gm, self.example_inputs, fw_compiler=ts_compile
-        )  # type: ignore[call-arg]
-
-
-aot_ts = AotTorchscript.compile_fn
+DEBUG = False
```
Comment (albanD) [inline torch/_dynamo/optimizations/training.py:154]: Should this be in the config?

File: torch/_dynamo/optimizations/training.py [hunk 4]
```diff
@@ -22,6 +31,62 @@
...
+def aot_autograd(**kwargs):
+    def compiler_fn(gm: torch.fx.GraphModule, example_inputs):
+        import functorch.compile
+
+        # Hack to get around circular import problems with aot_inductor_debug
+        if callable(kwargs.get("decompositions")):
+            kwargs["decompositions"] = kwargs["decompositions"]()
+
+        # TODO: stop monkeypatching here (without even cleaning up, UGH!)
+        functorch.compile.config.use_functionalize = True
```
Comment (albanD) [inline torch/_dynamo/optimizations/training.py:43]: Shouldn't we just update the config's default at this point?

**Review by albanD — APPROVED** [review]
Sounds good!

File: torch/_dynamo/optimizations/training.py (hunk 3 above)
Comment (ezyang) [reply torch/_dynamo/optimizations/training.py:154]: Yes, probably. Need to negotiate a name for it.

File: torch/_dynamo/optimizations/training.py (hunk 4 above)
Comment (ezyang) [reply torch/_dynamo/optimizations/training.py:43]: Yes, see https://github.com/pytorch/pytorch/pull/89663

File: torch/_dynamo/optimizations/training.py (hunk 2 above)
Comment (Chillee) [reply torch/_dynamo/optimizations/training.py:48]: Yeah I'm fine.

Comment (wconstab) [reply torch/_dynamo/optimizations/training.py:48]: is it truly pointless? what if the single op decomposes into some ops that can be fused?

Comment (ezyang) [reply torch/_dynamo/optimizations/training.py:48]: we gonna remove this conditional!

**Pull request closed without merging**